### ISSUE TITLE
Armature animation export problem

### ISSUE THREAD

**Issue by turanszkij** [KhronosGroup/glTF-Blender-IO#406]
Hello all, 

I am trying to use Blender 2.80 beta (latest as of today) to export a simple animated character to GLTF. It has only 1 mesh using 1 armature skinning animation and 1 mesh parented to a single bone. The skinning animation export worked correctly in Blender 2.79 using the old export script, but the bone parented object was incorrectly transformed, but now neither works in the new 2.80 version.
[girl_sword.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/3038929/girl_sword.zip)

I tried to load with the windows 10 model viewer, where the model shows up incorrectly animated and jittery and the bone parented object doesn't animate at all. Also tried with the Cesium gltf viewer, which fails to open the file: https://www.virtualgis.io/gltfviewer/ and an other online viewer which also fails: https://gltf-viewer.donmccurdy.com/
And in my own loader ( https://github.com/turanszkij/WickedEngine ) I get completely broken skinning animation but the bone parented object is also animated.

So to recap, the Blender 2.79 exporter is working kind of well (except bone parenting), but the same project in Blender 2.80 fails with bone parenting and skinning too.

**Comment by donmccurdy** [comment]
What result were you seeing on today's beta? The latest version of this repo (which is not yet in the Beta, I don't think) has several related fixes and is looking close...

![Screen Shot 2019-04-03 at 10 28 31 AM](https://user-images.githubusercontent.com/1848368/55499643-57d15680-55fb-11e9-8143-a97c39902baa.png)

... but the animation isn't playing quite correctly, there are some buggy cross-axis flips coming from somewhere. The bone parenting looks correct now.

**Comment by alexGuntha** [comment]
Hello,

I'm having a similar issue: the attached model is completely flat, but once exported, all animated joints are rotating in axes they shouldn't. Tested the file with Don McCurdy's gltf viewer. Using blender 2.79. I actually had the same issue with a previous version of the exporter, from this repository: https://github.com/KhronosGroup/glTF-Blender-Exporter , that's why I updated to glTF-Blender-IO today.
[bonhomme.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/3040473/bonhomme.zip)

**Comment by turanszkij** [comment]
I have made some GIFs to visualize the problem. I should mention, that I tried to copy the latest gltf exporter scripts to Blender 2.80 scripts/addons folder too, but the results are the same as yesterday.

This is the reference as it looks in Blender: 
![blender](https://user-images.githubusercontent.com/12954849/55549505-ee6e4800-56cd-11e9-86da-b7dcadef128b.gif)

This is how it looks in Windows 10 model viewer: 
![gltf_win10](https://user-images.githubusercontent.com/12954849/55549526-fa5a0a00-56cd-11e9-9231-c0b2c10836b6.gif)

This is how it looks in my gltf loader: 
![gltf_wickedengine](https://user-images.githubusercontent.com/12954849/55549552-08a82600-56ce-11e9-832d-087715648cd7.gif)

Notice that the Windows 10 viewer doesn't animate the "sword" that is parented to bone, but not skinned. The skinned model animation also looks wrong. The first frame geometry is completely blown up, and the last frame is completely collapsed. The intermediate frames are also wrong.
In my gltf loader, the first and last frames match the windows 10 loader if I am just eyeballing it, however, the intermediate frames are different (as in completely messed up), but the sword is animated too.

I understand that the windows 10 model viewer and my gltf loader are maybe not acceptable reference implementations, but they can generally handle every gltf model fine that I throw at it (eg. from Sketchfab).

**Comment by turanszkij** [comment]
New update: If I change the animation's interpolation mode to linear and re-export it, I get better results, but still not perfect. The blender gif in my previous post was actually using linear interpolation mode, but didn't re-export it with those settings. The legs are weird, and seems like only half of the animation is actually exported.

Windows 10 viewer:
![gltf_win10_2](https://user-images.githubusercontent.com/12954849/55552835-6f314200-56d6-11e9-9b61-4d530385b04f.gif)

My gltf loader:
![gltf_wickedengine_2](https://user-images.githubusercontent.com/12954849/55552851-79534080-56d6-11e9-9069-bf01e85ed0fb.gif)

**Comment by turanszkij** [comment]
When I try exporting a more complex character, the exporter crashes with this call stack: 

```
Traceback (most recent call last):
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\__init__.py", line 424, in execute
    return gltf2_blender_export.save(context, export_settings)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_export.py", line 40, in save
    json, buffer = __export(export_settings)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_export.py", line 54, in __export
    __gather_gltf(exporter, export_settings)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_export.py", line 69, in __gather_gltf
    scenes, animations = gltf2_blender_gather.gather_gltf2(export_settings)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather.py", line 36, in gather_gltf2
    animations += __gather_animations(blender_scene, export_settings)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather.py", line 62, in __gather_animations
    animations += gltf2_blender_gather_animations.gather_animations(blender_object, export_settings)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_animations.py", line 38, in gather_animations
    animation = __gather_animation(blender_action, blender_object, export_settings)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_animations.py", line 55, in __gather_animation
    channels=__gather_channels(blender_action, blender_object, export_settings),
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_animations.py", line 89, in __gather_channels
    blender_action, blender_object, export_settings)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_cache.py", line 51, in wrapper_cached
    result = func(*args)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_animation_channels.py", line 35, in gather_animation_channels
    channel = __gather_animation_channel(channel_group, blender_object, export_settings)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_animation_channels.py", line 52, in __gather_animation_channel
    sampler=__gather_sampler(channels, blender_object, export_settings),
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_animation_channels.py", line 85, in __gather_sampler
    export_settings
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_cache.py", line 51, in wrapper_cached
    result = func(*args)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_animation_samplers.py", line 41, in gather_animation_sampler
    input=__gather_input(channels, blender_object_if_armature, export_settings),
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_cache.py", line 51, in wrapper_cached
    result = func(*args)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_animation_samplers.py", line 71, in __gather_input
    export_settings)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_cache.py", line 51, in wrapper_cached
    result = func(*args)
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_animation_sampler_keyframes.py", line 174, in gather_keyframes
    for c in channels
  File "C:\Program Files\blender-2.80.0\2.80\scripts\addons\io_scene_gltf2\blender\exp\gltf2_blender_gather_animation_sampler_keyframes.py", line 174, in <listcomp>
    for c in channels
ZeroDivisionError: float division by zero

location: <unknown location>:-1

location: <unknown location>:-1

```
[Havoc.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/3079628/Havoc.zip)

The attached blender file has many animations both skinning and bone attached as well.

**Comment by AlbertoElias** [comment]
I´m getting errors as well when exporting animations:

![image](https://user-images.githubusercontent.com/913117/56473556-2ebf1b80-646d-11e9-93c4-4939e296c49a.png)

I've uploaded the zip. It works fine if I don't export animations.

[avatar.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/3101447/avatar.zip)

**Comment by julienduroure** [comment]
NLA issue is probably linked to #461 

I have to test again others cases, as some animation fixed was committed since the last comment

**Comment by julienduroure** [comment]
Everything seems to work now.
Don't hesitate to comment, open again this ticket, or create a new one if you found some (other) issue.

**Comment by alexGuntha** [comment]
[bonhomme.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/3246543/bonhomme.zip)

These are my last results:
=>In blender 2.79, animations are exported but rotations are slightly wrong (see bonhomme_blender2_79.glb)
=>In blender 2.8, I can see the animations' names in [https://gltf-viewer.donmccurdy.com/](url) but they don't play at all. (see bonhomme_blender2_8.glb)

The source .blend file is included.

**Comment by julienduroure** [comment]
@alexGuntha This seems to be a different problem, I will open a new issue about that.

Note that your armature object have some none default rotation and scale, that is always a very bad idea in Blender (don't know yet if the is related or not)

**Comment by julienduroure** [comment]
See #525
